### DISCUSSION TITLE
Meross MS600

### DISCUSSION THREAD

**Discussion by Taglar** [krahabb/meross_lan#485] [Ideas]
Hi, 

Meross has launched the MS600 motion sensor or presence sensor. It would be great if you could implement it.

It works great in the Meross app, but the possibilities are too limited.

Greetings,

Jürgen

**Comment by krahabb** [comment]
Hello @Taglar,
If you have the device installed (it should be anyway recognized by meross_lan) could you upload a 'Download diagnostic' for the device?

**Comment by Taglar** [comment]
Re,

sure, here it is:

[config_entry-meross_lan-01J74AB4XRBMDR99TZHZY39PB7.json](https://github.com/user-attachments/files/16922211/config_entry-meross_lan-01J74AB4XRBMDR99TZHZY39PB7.json)

Thx.

**Comment by Taglar** [comment]
Re,

here some Screenshots. If you need more, let me know.

Greets,

Jürgen

![Screenshot_20240909-112157](https://github.com/user-attachments/assets/1a0ed6a1-c66b-4d0c-b278-3b59244147dc)
![Screenshot_20240909-112208](https://github.com/user-attachments/assets/e459d857-846b-4139-9790-67954d186013)
![Screenshot_20240909-112219](https://github.com/user-attachments/assets/b61e7273-3933-499b-b07a-df2be6999de1)
![Screenshot_20240909-112239](https://github.com/user-attachments/assets/eab5f9f8-4b36-4981-a846-87be7817396a)
![Screenshot_20240909-112251](https://github.com/user-attachments/assets/3d28d5e7-cce0-46b3-ac04-09b12f1f05c1)

**Comment by krahabb** [comment]
The trace you've posted is really a good start. The device exposes some interesting data but I don't have a clue about their meanings.

There's one structure describing what should be the configuration of the device:
```json
{
  "config": [
    {
      "channel": 0,
      "mode": {
        "workMode": 2,
        "testMode": 2
      },
      "noBodyTime": {
        "time": 30
      },
      "distance": {
        "value": 12000
      },
      "sensitivity": {
        "level": 2
      },
      "mthx": {
        "mth1": 120,
        "mth2": 288,
        "mth3": 72
      }
    }
  ]
}
```

and then another message carrying the actual state:
```json
{
  "study": [
    {
      "channel": 0,
      "value": 2,
      "status": 1
    }
  ]
}
```

meross_lan, among its diagnostics features allows you to enable (and track) the so called 'diagnostic entities' i.e. whatever meross_lan doesn't understand gets reported as a sensor and updated from time to time. This feature can be enabled in the configuration of the device entry under the menu 'Diagnostics' -> 'Create diagnostic entities'

This way you can easily track in the device dashboard how the device state gets represented by the values in those 'config' and 'study' json payloads.

I'm only not totally sure these payloads can be parsed correctly since they're not plain dictionaries and the 'diagnostic code' in meross_lan might be set to discard lists/arrays since they usually carry huge amounts of data. I'll check this and follow-up.

**Comment by Taglar** [comment]
Re,

I don't know if I did it right. I activated debugging under Configuration, but apart from 2 restarts of the sensor, nothing happened.

I then started debugging manually on the left. The file is attached.

If you want to access the system yourself, please let me know.

[home-assistant_meross_lan_2024-09-09T19-21-05.776Z.log](https://github.com/user-attachments/files/16935549/home-assistant_meross_lan_2024-09-09T19-21-05.776Z.log)

![1](https://github.com/user-attachments/assets/2e0b6088-861e-4755-83c3-a39798dad7a5)
![2](https://github.com/user-attachments/assets/f24d9135-395f-46d7-9336-f73a2790a6c9)
![3](https://github.com/user-attachments/assets/981af427-95a6-4dfd-ad94-816696a29150)
![4](https://github.com/user-attachments/assets/eb525a6a-2b5e-46ec-b34d-4dfd4b081169)
![5](https://github.com/user-attachments/assets/28e98e58-d4d9-4997-89ec-6682e4e7a4a4)


Thx,

Jürgen

**Reply by krahabb** [reply]
Hello Jürgen,
The debug log shows a lot of useful messages though. I'll start implementing a basic behavior in the coming days so that, with the new version, you could do some deeper testing.

I think the relevant informations are carried in messages like (carried by namespace: `Appliance.Control.Sensor.LatestX`):
```json
    "payload": {
        "latest": [
            {
                "channel": 0,
                "data": {
                    "presence": [
                        {
                            "value": 2,
                            "timestamp": 1725908573,
                            "times": 0,
                            "distance": 3600
                        }
                    ]
                }
            }
        ]
    }
```
and:
```json
    "payload": {
        "latest": [
            {
                "channel": 0,
                "data": {
                    "light": [
                        {
                            "value": 22,
                            "timestamp": 1725908596
                        }
                    ]
                }
            }
        ]
    }
```

The message carrying the `light` info is pretty self-explanatory and the carried `value` should represent the light intensity in LUX.
The `presence` info is a bit more articulated: beside the `distance` (unit could be mm) of detection, the values reported by `value` and `times` are less clear to me. I see in the log you've posted that the `times` value keeps increasing but I have to better extract the logs and identify any pattern there.

For sure, the configuration you did is working as expected since you can now see those 'diagnostic entities' with 'Config...' and 'Study...'
Sadly enough, these other messages (on `Appliance.Control.Sensor.LatestX`) carrying more useful infos are not correctly parsed since there's a conflict with the syntax for the namespace which is expected to have a different layout (this comes from the implementation of another device using the same namespace ...)

**Comment by Taglar** [comment]
Re,

sorry my english is awful :)

"carrying more useful infos are not correctly parsed since there's a conflict with the syntax for the namespace which is expected to have a different layout (this comes from the implementation of another device using the same namespace ...)"

Does ist mean i should rename something - would not be a Problem.

Greets,

Jürgen

**Comment by ScottCLT** [comment]
I got 2 of these not realizing they weren't yet compatible with home assistant. I've setup automations in Google Home, but they are way too slow. Walk in a room, wait a few seconds, and the light turns on. Would be much faster doing it all locally in HA. I will keep up to date on this project and hopefully the device will be supported soon.

**Comment by krahabb** [comment]
An initial implementation is under way but I'm very busy atm so expect  some days ..maybe a week

**Reply by Taglar** [reply]
Sorry that i write not much atm.
This week I'm away on business and then on vacation for 2 weeks.

**Comment by Taglar** [comment]
Hi,

take your time - I'm happy that you implement meross what I can't do.

Thanks,

Jürgen

**Comment by ScottCLT** [comment]
Agreed - take your time. I have it working now, outside of HA, through Google Home, so it's still usable. It's just slow because of the need for cloud communication. Also, Google Home doesn't allow me to set active times for the automation to be running (I only want to run in at night when it's dark outside. I can do that in HA, but not Google. Nonetheless, it's mostly working how I want (just a bit slowly). We appreciate your efforts for doing what most of us aren't capable of doing. We can wait a bit longer. 😁 Thank you!

**Reply by Taglar** [reply]
Re,

I have regulated this with a meross power strip. 

In the automation I set that the slot is switched on in the morning and off in the evening.  Normally, this should also work with sunset and sunrise. 

Cu

**Comment by krahabb** [comment]
The new (preview/beta) release 5.4.0 has some support now. It should expose the configuration and the 'presence' data.
Sensor names are a bit raw at the moment so I will eventually ask you to 'match' those names to actual names in the app so that I can refine those. 
The configuration entities are supposed to work i.e. receive current conf and also send data to the device but I don't know if this is for sure. I'm waiting your experiments;).

Also, the 'select' entities are now prepopulated with basic values and meross_lan should automatically add any other raw value it sees from the device for the same entity but again, the meaning of these 0 - 1 - 2 - etc are not known and need to be mapped to actual meaningful labels (for example the 'workMode' select entity should map those numbers to the standard nomenclature used in app - same for every other)

The configuration entities are now either `number` or `select` and I've choosen numbers where I suppose there's a full range of values to be configured (for example the 'noBodyTime') while select are those that I expect to only represent a limited amount of options/values where each of these values has a semantically different meaning.

As for the real 'presence' detection, the protocol carries 3 values in the related message and I've mapped each of these to its own (sensor) entity.
- Presence: I guess this is an 'estimate' of the likelihood of presence with higher values representing higher probability of presence 
- Presence distance: should clearly be the distance of the subject from the sensor
- Presence times: I really don't know...in the logs I see values going up and somewhat down here and there but I can't really figure it out

There's also the `Light` entity which should carry its own values.

I'll wait for your feedbacks, both comments and some diagnostic to refine the job...no rush anyway

**Reply by Taglar** [reply]
Thanks, i updated to the alpha.
I think i can give some answers - i will test it later at home.

noBodyTime is the time, after which “absent” is marked if no movement is detected.

So if I leave the room, the lights go out after 30 seconds due to automation, as the status changes to absent.

Distance ist from min 0.5 to max 12 meters:

![grafik](https://github.com/user-attachments/assets/007fe943-6ee3-4564-8ff2-5f9399159f71)

**Automatic translation:**

0-6 Meter 

The sensor detects the user's presence and it is recommended to position the sensor as close to the chest as possible during use. 

6-12 Meter 

The sensor detects subtle movements of the user, and as the distance increases, larger movements are required. 

If the maximum detection distance is exceeded, the sensor will not trigger. Please test the range and the size of the movement in the 'Calibrate installation'. 

Greets,

Jürgen

**Comment by ScottCLT** [comment]
Perhaps this is outside of this discussion, but I just installed the alpha today and I can now see my Meross MS600 presence sensor send a value of "1" when the room is empty and "2" when someone is present (and sometimes it changes to 6 just for a second before changing back to 2). Once you exit the room, the state goes back to "1" after 15 seconds (how I have it configured in the Meross app. This seems like a great initial attempt at this integration! Thank you, community!!!

The one thing I'm having a problem with is configuring triggers for automations. Do I choose the device state (the presence sensor device) to detect the state "present" (didn't work for me). I also tried seeing if the entity of the presence sensor changing values from 1 to 2 would work, but that didn't trigger the automation either. 

I know this is still in its early stages, but what should be used to trigger an automation upon present/not present state?

Thanks a bunch!

**Reply by krahabb** [reply]
Hello @ScottCLT,
Thank you for your reply. This will help refining the coding of the presence status to more meaningful labels.

Regarding the value '6' you temporarily see, could it be that `6` means 'movement' while `2` means (rather stationary) 'presence' ?

As for the automation I think you should better check the composition of that. Ensure you are using using the correct entity i.e. the 'Presence' sensor. Your approach seems correct though, either checking a state change or matching a state value. The problem might be the data format you use..sometimes it is not very clear in HA (to me at least) if a number is represented as a number or as a string so you might just need to add some `""` enclosing to the values you provide to check the state (or remove them..).

The 'Presence' entity is designed to be a numeric entity in `meross_lan` but since I didn't provide any `device_class` or `unit of measurement` HA might think it is a 'plain-old-style' string-valued sensor entity.

**Reply by ScottCLT** [reply]
Thanks for the response, @krahabb. I test putting double quotes around the value, but the automation is still not running when the number changes.

I tested the sensor status values. It's 1.0 for not present and 2.0 for present. I can't get it to register a 6.0 anymore, even after doing a little dance in front of it / it remained at 2.0. I actually haven't had a (monitored) change to 6.0 in white.

I think I am going to try the automation with a trigger of 1.0 instead of just 1 like it is now. I will update after I've accomplished that.

**Reply by ScottCLT** [reply]
I just tested with a presence value of "2.0" and 2.0 (I.e with and without the double quotes) and had the same problem. I see the value on the device change in HA when present (2.0), but still couldn't get it register to trigger an automation.

Thanks again those working on this project. I would be participating in the development, myself, but I lack the development skills.

**Reply by ScottCLT** [reply]
I've got this automation working using scene. Presence = one scene and no presence= another scene. It works! My only complaint is with the speed. It takes a full second at least to activate to automation. Does anyone know how to make it quicker?

**Reply by krahabb** [reply]
I don't know if scenes in HA are so slow but I guess part of the delay might be due to the nature of cloud MQTT so that the device message (presence/not presence) has to go out to a cloud MQTT broker and then HA will receive the notification so it is quite a rountrip..

**Reply by ScottCLT** [reply]
The notification for triggering the automation was delayed. After that the light came right on, so I assume it's inherent to the sensor. Is there any way to make this integration local?

**Reply by krahabb** [reply]
I think 1 second is rasonable by the fact meross_lan get asynchronous updates from your device through cloud MQTT. In order to reduce latency you should 'pair' your device to your private (local) MQTT broker if any.. here the delays are in the order of milliseconds but this would mean the Meross app would likely not work anymore to control the device.

Another option would be to try the 'Matter integration' in HA and see if this device is nicely handled/supported there. With Matter, which is a local protocol everything should be almost instant.

**Comment by Taglar** [comment]
I'm back.

What should i test and what information do you need?

I noticed that I cannot select the presence sensor as a motion detector via Blueprint.

![Motion](https://github.com/user-attachments/assets/c530511b-edf4-4c5d-8716-2d331378bf6e)
![BluePrint](https://github.com/user-attachments/assets/706f2a95-f63a-4919-b085-d1816a89354d)

**Reply by krahabb** [reply]
The problem is the blueprint only accepts binary sensor entities with 'Motion' device class i.e. entities that just switch between 'motion detected' and 'Clear'. You could build a template binary sensor using the state of the meross_lan presence sensor for that so that you'd be able to use the blueprint..else manually implement the same automation based off the state of the meross_lan sensor.
Development for building that specific binary sensor type is not that hard..just I'm lacking time these days..
I'll see if I can push that together with a fix for a compatibility bug with HA core 2024.10 (#497)

**Reply by Taglar** [reply]
No Stress. Would it help if you have one Sensor? If yes, how can I buy u a Coffee?

**Reply by krahabb** [reply]
Hello @Taglar,
I've managed to 'sneak-in' the MOTION binary sensor to the official 5.4 release..I'll wait your feedback about it fixing (or not) the issue

**Reply by Taglar** [reply]
Hi,

the sensor can now be found - thx.

![SensorBluePrint](https://github.com/user-attachments/assets/813675e0-bfd9-4c3b-9085-369678cac10d)

I integrated one in HA with matter.

**Here are the Logs with a short test:**

[home-assistant_matter_2024-10-19T20-51-38.797Z.log](https://github.com/user-attachments/files/17447349/home-assistant_matter_2024-10-19T20-51-38.797Z.log)

**Here is the Screenshot from the protocol in the Meross App:**

![Prot](https://github.com/user-attachments/assets/d09dfd82-2f4e-4691-aa0d-ced6f70af480)
![Lux1](https://github.com/user-attachments/assets/578e0d00-7cd7-4168-8ab0-7eb06dfbc945)
![Lux2](https://github.com/user-attachments/assets/c4858724-b914-48af-b429-af6cbd6a5f92)

**Dashboard with Matter:**

![MatterInt](https://github.com/user-attachments/assets/16987e4a-7cae-4f57-9455-3197de4797d5)

**Dashboard with your Plugin:**

![HAInt](https://github.com/user-attachments/assets/f8f05dd2-2c34-4901-9509-320c9762acd2)

If i press the button (Drücken), the Sensor is triggered:

![Matter log](https://github.com/user-attachments/assets/04138a77-f17a-47c4-94a2-617540835be6)

(Last three entries - not included in the matter Logs)

**Comment by Lestat78** [comment]
Hi @krahabb,
is there any way I can support the developments for the MS600 by donating a couple of those devices to you? I see you are in Italy, same as me, so I would gladly send you an Amazon Italy gift card, if you like.
I have many Meross devices and one of the reasons I got them was because of your integration with Home Assistant, and with the black friday sales I have purchased a few MS600 to drop around the house, so I'd really like to somehow contribute to the great work you have done!

Thanks :)

**Comment by krahabb** [comment]
Hello @Lestat78,
Thank you for your support but I cannot really put much time into meross_lan at the moment so I'm in almost 'maintanence mode' with the project (even if I'm not really going to completely drop development/support for this since I still really love it).
I appreciate your offer but for the time being I would 'pass' at least until I free up some time to invest into this project again.

Nevertheless, actual support/implementation for MS600, even though still very basic, should be enough to succesfully integrate it into HA.
Is there any 'missing' feature that you would like to add? Sometimes, a few diagnostics/traces together with user tests are enough to refine or add behaviors to the component.

**Comment by sphinxlady** [comment]
Polling for states is still broken but it appears to be on meross end not home/meross lan.
Even with the phone app you have to refresh it to get any kind of updated state change,

I suspect this has been done on purpose to reduce overhead on their part as things move to matter and the like.

Honestly the matter connections local anyway and the app still gives you total control of its setting so the polling issues in this instance it minor.
